PowerEditor/src/WinControls/ProjectPanel/ProjectPanel.cpp

```diff
@@ -1208,10 +1787,246 @@ int FileRelocalizerDlg::doDialog(const TCHAR *fn, bool isRTL)
+	{
+		::SendMessage(hCombo, CB_DELETESTRING, i, 0);
+	}
+
+	i = ::SendMessage(hCombo, CB_INSERTSTRING, 0, reinterpret_cast<LPARAM>(txt2add));
+
+	::SendMessage(hCombo, CB_SETCURSEL, i, 0);
+}
+
+generic_string FilterDlg::getTextFromCombo(HWND hCombo, bool) const
+{
+	TCHAR str[PROJECTPANEL_FILTERS_MAXLENGTH];
+	::SendMessage(hCombo, WM_GETTEXT, PROJECTPANEL_FILTERS_MAXLENGTH - 1, reinterpret_cast<LPARAM>(str));
+	return generic_string(str);
+}
+
+
+
+void FilterDlg::fillComboHistory(int id, const std::vector<generic_string> & strings)
+{
+	HWND hCombo = ::GetDlgItem(_hSelf, id);
+
+	for (std::vector<generic_string>::const_reverse_iterator i = strings.rbegin(); i != strings.rend(); ++i)
+	{
+		addText2Combo(i->c_str(), hCombo);
+	}
+	::SendMessage(hCombo, CB_SETCURSEL, 0, 0); // select first item
+}
+
+int FilterDlg::saveComboHistory(int id, int maxcount, std::vector<generic_string> & strings)
+{
+	TCHAR text[PROJECTPANEL_FILTERS_MAXLENGTH];
+	HWND hCombo = ::GetDlgItem(_hSelf, id);
+	int count = ::SendMessage(hCombo, CB_GETCOUNT, 0, 0);
+	count = min(count, maxcount);
+
+	if (count == CB_ERR) return 0;
+
+	if (count)
+		strings.clear();
+
+	for (int i = 0; i < count; ++i)
+	{
+		::SendMessage(hCombo, CB_GETLBTEXT, i, reinterpret_cast<LPARAM>(text));
+		strings.emplace_back(generic_string(text));
+	}
+	return count;
+}
+
+
+int FilterDlg::doDialog(bool isRTL /*= false*/)
+{
+	if (isRTL)
+	{
+		DLGTEMPLATE *pMyDlgTemplate = NULL;
+		HGLOBAL hMyDlgTemplate = makeRTLResource(IDD_PROJECTPANEL_FILTERDIALOG, &pMyDlgTemplate);
+		int result = ::DialogBoxIndirectParam(_hInst, pMyDlgTemplate, _hParent,  dlgProc, reinterpret_cast<LPARAM>(this));
+		::GlobalFree(hMyDlgTemplate);
+		return result;
+	}
+
+
+	return ::DialogBoxParam(_hInst, MAKEINTRESOURCE(IDD_PROJECTPANEL_FILTERDIALOG), _hParent,  dlgProc, reinterpret_cast<LPARAM>(this));
+
+}
+
+ProjectPanel::TreeUpdaterDirectory::TreeUpdaterDirectory(ProjectPanel* projectPanel, HTREEITEM hItem) 
+	: Directory()
+	, _projectPanel(projectPanel)
+	, _treeView(&projectPanel->getTreeView())
+	, _hItem(hItem)
+{
+	TCHAR textBuffer[MAX_PATH];
+	TVITEM tvItem;
+	tvItem.mask = TVIF_TEXT | TVIF_PARAM;
+	tvItem.pszText = textBuffer;
+	tvItem.cchTextMax = MAX_PATH;
+
+	tvItem.hItem = hItem;
+	SendMessage(_treeView->getHSelf(), TVM_GETITEM, 0,reinterpret_cast<LPARAM>(&tvItem));
+	ProjectPanelData& data = *reinterpret_cast<ProjectPanelData*>(tvItem.lParam);
+	assert (data.isBaseDirectory() || data.isDirectory());
+	if (!data.isBaseDirectory() && !data.isDirectory())
+		return;
+	_exists = PathFileExists(data._filePath.c_str()) != 0;
+	if (_exists)
+		_path = data._filePath;
+
+    for (HTREEITEM hItemNode = _treeView->getChildFrom(hItem);
+		hItemNode != NULL;
+		hItemNode = _treeView->getNextSibling(hItemNode))
+	{
+		tvItem.hItem = hItemNode;
+		SendMessage(_treeView->getHSelf(), TVM_GETITEM, 0,reinterpret_cast<LPARAM>(&tvItem));
+		ProjectPanelData& data = *reinterpret_cast<ProjectPanelData*>(tvItem.lParam);

```

Comment:
You ***absolutely*** should check to see if `::SendMessage` returned `FALSE` before using `tvItem.lParam`.